### ISSUE TITLE
Add file name information to location info

### ISSUE THREAD

**Issue by dmajda** [pegjs/pegjs#421]
It is sometimes useful to include a file name in location info. For example, when implementing a language with a module system, one may want to display syntax errors coming from a PEG.js-generated parser with the module name included. Similarly, after parsing, one may want to do a semantic checks on AST nodes produced by the parser and again display them with the module name included.

Right now, the above is possible by tracking filenames externally (so the program would know which file it is parsing or checking). Alternatively, one can do what I did in [Juttle](https://github.com/juttle/juttle) and manually adjust location info [in the parser](https://github.com/juttle/juttle/blob/163fb4f62af76dc3b33c05b45e7552b865b9900c/lib/parser/parser.pegjs#L27-L29) and [in thrown exceptions](https://github.com/juttle/juttle/blob/163fb4f62af76dc3b33c05b45e7552b865b9900c/lib/parser/index.js#L227). Neither of this is super-convenient.

The proposal is to add a new parser option, `filename`. If set, the `filename` property in location info returned by the `location` function or in thrown exceptions would be set to its value. If not set, the `filename` property would be set to `null`.

This issue originated in #402.

**Comment by matthew-dean** [comment]
Great idea.

**Comment by futagoza** [comment]
I achieve this result in my generated parsers by always passing a `filename` in the options and then adding this to the top of my PEG.js grammar as initializer code:

```js
var pegjs$location = location

location = function getLocation() {
  var details = pegjs$location()
  details.filename = options.filename
  return details
}
```

### my suggestion...

Rather then implementing this in generated parsers, I suggest you either:

1) leave it as it is right now
2) make `location()` only return `[startPos, endPos]`

By leaving it as it is you encourage the parser developer to experiment with PEG.js features and let them find what works best for them, feature or performance wise.

If the generated parsers `location()` helper only returns an array containing the `startPos` and `endPos`, then this increase's overall performance of the parser as right now every time we use `location()` it calculates the line and column for both `startPos` and `endPos`, and if your developing a parser on the scale of a ECMAScript like language, this can be quite a hit.

The parser developer can manually implement their own location function similar to mine above to achieve the same result if they wish or using generated parser internals get their own results that depending on what they do can either perform better or be more feature rich then the current `location()` helper.

**Comment by futagoza** [comment]
In relation to the suggestions I made above, point 2 was implemented by @felix9 in #528 with `range()`.

I will soon add a commit that simply checks at compile time the option `features` (default value is `{ filename: true, ...helpers: true }`), where `helpers` denotes the names of the builtin helpers (e.g. `location: true`, `range: true`, `etc`) and `filename` indicates that the location helper should add the filename to the result if passed to the generated parser. Adding this option will also allow in the future to expand the range of options available to the code generator.
